site/src/pages/WorkspacesPage/data.ts

```diff
@@ -0,0 +1,113 @@
+import {
+  QueryKey,
+  useMutation,
+  useQuery,
+  useQueryClient,
+} from "@tanstack/react-query"
+import { getWorkspaces, updateWorkspaceVersion } from "api/api"
+import {
+  Workspace,
+  WorkspaceBuild,
+  WorkspacesResponse,
+} from "api/typesGenerated"
+
+const getQueryKey = (page: number, query: string) => {
+  return ["workspaces", query, page]
+}
+
+type UseWorkspacesDataParams = {
+  page: number
+  limit: number
+  query: string
+}
+
+export const useWorkspacesData = ({
+  page,
+  limit,
+  query,
+}: UseWorkspacesDataParams) => {
+  const queryKey = getQueryKey(page, query)
+  const result = useQuery({
+    queryKey,
+    queryFn: () =>
+      getWorkspaces({
+        q: query,
+        limit: limit,
+        offset: page,
+      }),
+    refetchInterval: 5_000,
+  })
+
+  return {
+    ...result,
+    queryKey,
+  }
+}
+
+export const useWorkspaceUpdate = (queryKey: QueryKey) => {
+  const queryClient = useQueryClient()
+
+  return useMutation({
+    mutationFn: updateWorkspaceVersion,
+    onMutate: async (workspace) => {
+      await queryClient.cancelQueries({ queryKey })
+      queryClient.setQueryData<WorkspacesResponse>(queryKey, (oldResponse) => {
+        if (oldResponse) {
+          return assignPendingStatus(oldResponse, workspace)
+        }
+      })
+    },
+    onSuccess: (workspaceBuild) => {
+      queryClient.setQueryData<WorkspacesResponse>(queryKey, (oldResponse) => {
+        if (oldResponse) {
+          return assignLatestBuild(oldResponse, workspaceBuild)
+        }
+      })
```

Comment:
Do we need an `onError` key?

Reply:
Ahhh, good catch. I forgot to handle the update error. Thank you! But as the `useQuery`, the `useMutation` also returns an error object so if you want to just get the error and display it in the UI no `onError` is needed. 

```tsx
const updateVersion = useMutation()

if(updateVersion.error) {
   return <Alert error={error}>
}
```

but in this case, I think we can use `onError` to trigger a `displayError` notification.